File: app/src/main/java/ericdiaz/program/currencyconveterlive2019/repository/ExchangeRateRepository.java
package ericdiaz.program.currencyconveterlive2019.repository;

import androidx.annotation.NonNull;

import javax.inject.Inject;

import ericdiaz.program.currencyconveterlive2019.model.ExchangeRateResponse;
import ericdiaz.program.currencyconveterlive2019.network.ExchangeRateService;
import io.reactivex.Single;

public class ExchangeRateRepository implements BaseRepository {

    private final ExchangeRateService exchangeRateService;

    @Inject
    public ExchangeRateRepository(ExchangeRateService exchangeRateService) {
        this.exchangeRateService = exchangeRateService;
    }

    @Override
    public Single<ExchangeRateResponse> requestExchangeRates(@NonNull final String date,
                                                             @NonNull final String baseCurrency) {
        return exchangeRateService.getExchangeRates(date, baseCurrency);
    }
}

File: data/src/main/java/ericdiaz/program/data/repository/ExchangeRateRepository.kt
package ericdiaz.program.data.repository

import ericdiaz.program.data.ExchangeRatesQueries
import ericdiaz.program.data.db.ExchangeRateDatabase
import ericdiaz.program.data.model.ExchangeRateResponse
import javax.inject.Inject

class ExchangeRateRepository @Inject constructor(exchangeRateDatabase: ExchangeRateDatabase) : BaseRepository {

    private val exchangeRatesDatabaseQueries: ExchangeRatesQueries = exchangeRateDatabase.exchangeRatesQueries

    fun insertExchangeRateResponse(exchangeRateResponse: ExchangeRateResponse) {
        exchangeRatesDatabaseQueries
                .insertExchangeRates(
                        exchangeRateResponse.date,
                        exchangeRateResponse.ratesMap,
                        exchangeRateResponse.baseCurrency
                )
    }

    fun getCachedRatesByDateAndBaseCurrency(date: String,
                                            baseCurrency: String): ExchangeRateResponse {
        return exchangeRatesDatabaseQueries
                .selectByDateAndBaseCurrency(
                        date,
                        baseCurrency,
                        mapper = { cachedDate, cachedRateMap, cachedBaseCurrency ->
                            ExchangeRateResponse(
                                    date = cachedDate,
                                    ratesMap = cachedRateMap,
                                    baseCurrency = cachedBaseCurrency)
                        }
                ).executeAsOne()

    }

}